server/registry/internal/storage/list.go

```diff
@@ -25,10 +26,16 @@ import (
 	"gorm.io/gorm"
 )
 
-// ProjectList contains a page of project resources.
-type ProjectList struct {
-	Projects []models.Project
-	Token    string
+// limit returns the database page size to use for a listing request.
+func limit(opts PageOptions) int {
+	// Without filters, read exactly enough rows to fill the page,
+	// plus an extra row to check if another page exists.
+	if opts.Filter == "" {
+		return int(opts.Size) + 1
+	}
+
+	// When filters are present, we should read larger pages.
+	return 500
```

Comment:
In a simulation of a practical ListApis request (page_size=`10`, filter=`api_id == 'my-api'`), having this case results in ~64ms response time on my local machine, as opposed to ~187ms when the database always uses `page_size + 1` to choose its database page size.

Reply:
What (if anything) do we guarantee about the page sizes that get returned when filters are present? Is the "500" here because we only make one fetch per request?... so if none of those results pass the filter, could we return an empty page?

Reply:
I'm not totally sure I understand your question, so here's a full explanation of why this function exists.

When a database page is loaded, the list methods loop through it and ignore any rows that don't match the filter. If the response isn't filled up to the requested size, then another page is loaded from the database. This repeats until the response is full or the table has nothing left to read.

Without filters, we can load exactly the resources we need in one query. On the other hand, in cases where a filter only matches one row (or a few) in the entire table, we usually need to do a full table scan with `ceil(table_size/page_size)` queries to find all the matches. This can be a very large number of queries if the page size is too small on a large table. 

This `limit` function lets us compute a "smart" database page size for each request. As long as it returns positive integers, the listing methods will work correctly. The choice of 500 for filtered requests is somewhat arbitrary, aside from us wanting it to be large enough to avoid making too many database queries per request.